README.txt
= GChart

== DESCRIPTION
  
GChart exposes the Google Chart API (http://code.google.com/apis/chart) via
a friendly Ruby interface. It can generate the URL for a given chart
(for webpage use), or download the generated PNG (for offline use).

== PROBLEMS/TODO

* Add support fills (area or background), grid lines, shape markers, range markers 
* Support shorthand colors and color names
* Make venn data specification friendlier

There are lots of missing features. Until they're implemented, you can directly specify
query parameters using the :extras key, e.g.,

  # provides a legend for each data set
  g = GChart.line(:data => [[1, 2], [3, 4]], :extras => { "chdl" => ["First", "Second"] })

== SYNOPSIS

  # line chart
  g = GChart.line(:data => [0, 10, 100])
  
  # bar chart
  g = GChart.bar(:data => [100, 1000, 10000])
  
  # pie chart (pie3d for a fancier look)
  g = GChart.pie(:data => [33, 33, 34])
  
  # venn diagram (asize, bsize, csize, ab%, bc%, ca%, abc%)
  g = GChart.venn(:data => [100, 80, 60, 30, 30, 30, 10])
  
  # scatter plot (x coords, y coords [, sizes])
  g = GChart.scatter(:data => [[1, 2, 3, 4, 5], [5, 4, 3, 2, 1], [1, 2, 3, 4, 5]])
  
  # chart title
  g = GChart.line(:title => "Awesomeness over Time", :data => [0, 10, 100])

  # data set legend
  g = GChart.line(:data => [[1, 2], [3, 4]], :legend => ["Monkeys", "Ferrets"])

  # data set colors
  g = GChart.line(:data => [[0, 10, 100], [100, 10, 0]], :colors => ["ff0000", "0000ff"])
  
  g.to_url            # generate the chart's URL, or
  g.fetch             # get the bytes, or
  g.write("foo.png")  # write to a file (defaults to "chart.png")
  g.write(stream)     # write to anything that quacks like IO

== LICENSE

(The MIT License)

Copyright 2007 John Barnette (jbarnette@rubyforge.org)

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
2007-12-06  John Barnette <jbarnette@rubyforge.org>

  * Initial version.


spec/gchart_spec.rb
require File.expand_path("#{File.dirname(__FILE__)}/helper")

describe GChart do
  it "supplies a version" do
    GChart::VERSION.should_not be_nil
  end
end

describe GChart, ".encode" do
  it "only supports the extended encoding" do
    lambda { GChart.encode(:simple, 4, 10) }.should raise_error(ArgumentError)
  end

  it "implements the extended encoding" do
    expected = {
      0 => "AA", 25 => "AZ", 26 => "Aa", 51 => "Az", 52 => "A0", 61 => "A9", 62 => "A-", 63 => "A.",
      64 => "BA", 89 => "BZ", 90 => "Ba", 115 => "Bz", 116 => "B0", 125 => "B9", 126 => "B-", 127 => "B.",
      4032 => ".A", 4057 => ".Z", 4058 => ".a", 4083 => ".z", 4084 => ".0", 4093 => ".9", 4094 => ".-", 4095 => ".."
    }
    
    expected.each do |original, encoded|
      GChart.encode(:extended, original, 4095).should == encoded
    end
  end
  
  it "encodes nil correctly" do
    GChart.encode(:extended, nil, 1).should == "__"
  end
end

describe GChart, ".line" do
  it "can render a basic line URL" do
    expected = { "cht" => "lc", "chs" => "300x200", "chd" => "e:AAAo..", "chtt" => "test" }
    chart = GChart.line(:title => "test", :data => [1, 100, 10000])

    chart.query_params.should == expected
    chart.to_url.should == "http://chart.apis.google.com/chart?chs=300x200&cht=lc&chtt=test&chd=e:AAAo.."
  end
end


test/test_gchart.rb
require "test/unit"
require "gchart"

class GChart
  public :google_query_params, :google_chart_type, :google_data
end

class TestGChart < Test::Unit::TestCase
  def test_supplies_a_version
    assert_not_nil(GChart::VERSION)
  end
  
  def test_allows_hash_initialization
    assert_equal(:line, GChart.new(:type => :line).type)
  end
  
  def test_complains_about_unknown_options
    assert_raise(NoMethodError) { GChart.new(:monkey => :chimchim) }
  end
  
  def test_allows_block_initialization
    c = GChart.new do |chart|
      chart.type = :line
    end
    
    assert_equal(:line, c.type)
  end
  
  def test_allows_all_valid_chart_types
    GChart::TYPES.each do |type|
      assert_nothing_raised(ArgumentError) { GChart.new(:type => type) }
    end
  end
  
  def test_complains_about_invalid_chart_types
    assert_raise(ArgumentError) { GChart.new(:type => :monkey) }
  end
  
  def test_allows_size_to_be_specified_combined
    c = GChart.new(:size => "13x14")
    assert_equal(13, c.width)
    assert_equal(14, c.height)
  end
  
  def test_defaults
    c = GChart.new
    assert_equal(:line, c.type)   
    assert_equal([], c.data)
    assert_equal(300, c.width)
    assert_equal(200, c.height)
    assert(!c.horizontal?)
    assert(!c.grouped?)
  end
  
  def test_google_chart_types
    assert_equal("lc", GChart.new(:type => :line).google_chart_type)
    assert_equal("lxy", GChart.new(:type => :linexy).google_chart_type)
    
    assert_equal("bhg", GChart.new(:type => :bar, :horizontal => true, :grouped => true).google_chart_type)
    assert_equal("bhs", GChart.new(:type => :bar, :horizontal => true, :grouped => false).google_chart_type)
    assert_equal("bvs", GChart.new(:type => :bar, :horizontal => false, :grouped => false).google_chart_type)
    
    assert_equal("p", GChart.new(:type => :pie).google_chart_type)
    assert_equal("v", GChart.new(:type => :venn).google_chart_type)
    assert_equal("s", GChart.new(:type => :scatter).google_chart_type)
  end
  
  def test_google_extended_notation
    expected = {
      0 => "AA", 25 => "AZ", 26 => "Aa", 51 => "Az", 52 => "A0", 61 => "A9", 62 => "A-", 63 => "A.",
      64 => "BA", 89 => "BZ", 90 => "Ba", 115 => "Bz", 116 => "B0", 125 => "B9", 126 => "B-", 127 => "B.",
      4032 => ".A", 4057 => ".Z", 4058 => ".a", 4083 => ".z", 4084 => ".0", 4093 => ".9", 4094 => ".-", 4095 => ".."
    }
    
    expected.each do |original, encoded|
      assert_equal(encoded, GChart.encode_extended(original))
    end
    
    assert_equal("__", GChart.encode_extended(nil))    
  end
  
  def test_allows_extra_query_params
    c = GChart.new(:extras => { :foo => :bar })
    assert(c.google_query_params.include?(:foo))
  end
  
  def test_supports_title
    chart = GChart.new
    
    assert_nil(chart.google_query_params["chtt"])
    
    chart.title = "foo"
    assert_equal("foo", chart.google_query_params["chtt"])
    
    chart.title = "a space"
    assert_equal("a+space", chart.google_query_params["chtt"])
    
    chart.title = "a\nnewline"
    assert_equal("a|newline", chart.google_query_params["chtt"])
  end
  
  def test_supports_colors
    chart = GChart.new(:colors => ["cccccc", "eeeeee"])
    assert_equal(chart.google_query_params["chco"], "cccccc,eeeeee")
  end
  
  def test_generates_correct_query_params_and_url_for_simple_example
    expected = { "cht" => "lc", "chs" => "300x200", "chd" => "e:AAAo..", "chtt" => "test" }
    
    chart = GChart.new(:title => "test", :data => [1, 100, 10000])
    assert_equal(expected, chart.google_query_params)
    assert_equal("http://chart.apis.google.com/chart?chs=300x200&cht=lc&chtt=test&chd=e:AAAo..", chart.to_url)
  end
end


spec/gchart/bar_spec.rb
require File.expand_path("#{File.dirname(__FILE__)}/../helper")

describe GChart::Bar do
  before(:each) do
    @chart = GChart::Bar.new
  end
  
  it "is ungrouped by default" do
    @chart.grouped.should == false
    @chart.should_not be_grouped
  end
  
  it "is horizontal by default" do
    @chart.orientation.should == :horizontal
    @chart.should be_horizontal
    @chart.should_not be_vertical
  end
  
  it "renders the correct chart type" do
    GChart::Bar.new.render_chart_type.should == "bhs"
    GChart::Bar.new(:grouped => true).render_chart_type.should == "bhg"
    GChart::Bar.new(:orientation => :vertical).render_chart_type.should == "bvs"
  end
end


lib/gchart.rb
require File.expand_path(File.dirname(__FILE__) + "/version")

%w(base bar line pie pie_3d scatter venn xy_line).each do |type|
  require File.expand_path(File.dirname(__FILE__) + "/gchart/#{type}")
end

module GChart
  URL   = "http://chart.apis.google.com/chart"
  CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-.".split("")
  PAIRS = CHARS.collect { |first| CHARS.collect { |second| first + second } }.flatten

  class << self
    def line(*args, &block); Line.new(*args, &block) end
    def xyline(*args, &block); XYLine.new(*args, &block) end
    def bar(*args, &block); Bar.new(*args, &block) end    
    def pie(*args, &block); Pie.new(*args, &block) end    
    def pie3d(*args, &block); Pie3D.new(*args, &block) end    
    def venn(*args, &block); Venn.new(*args, &block) end    
    def scatter(*args, &block); Scatter.new(*args, &block) end
    
    def encode(encoding, n, max)
      unless encoding == :extended
        raise ArgumentError, "unsupported encoding: #{encoding.inspect}"
      end

      return "__" if n.nil?
      PAIRS[(n * (PAIRS.size - 1) / max).to_i]
    end
  end
end


lib/gchart/bar.rb
module GChart
  class Bar < GChart::Base
    ORIENTATIONS = [:horizontal, :vertical]
    
    attr_accessor :grouped
    alias_method :grouped?, :grouped
    
    attr_reader :orientation
    
    def initialize(*args, &block)
      @grouped = false
      @orientation = :horizontal
      super
    end
    
    def orientation=(orientation)
      unless ORIENTATIONS.include?(orientation)
        raise ArgumentError, "Invalid orientation: #{orientation.inspect}. " +
          "Valid orientations: #{ORIENTATIONS.inspect}"
      end
      
      @orientation = orientation
    end
    
    def horizontal?
      @orientation == :horizontal
    end

    def vertical?
      @orientation == :vertical
    end
    
    def render_chart_type #:nodoc:
      "b#{@orientation.to_s[0..0]}#{grouped? ? "g" : "s"}"
    end
  end
end


lib/gchart/base.rb
require "open-uri"
require "uri"

module GChart
  class Base
    attr_accessor :data
    attr_accessor :extras
    attr_accessor :title
    attr_accessor :colors
    attr_accessor :legend
    attr_accessor :max

    attr_reader :width    
    attr_reader :height

    def initialize(options={}, &block)
      @data = []
      @extras = {}
      @width = 300
      @height = 200
  
      options.each { |k, v| send("#{k}=", v) }
      yield(self) if block_given?
    end

    # Sets the chart's width, in pixels. Raises +ArgumentError+ if +width+ is less than 1.
    def width=(width)
      raise ArgumentError, "Invalid width: #{width.inspect}" if width.nil? || width < 1
      @width = width
    end

    # Sets the chart's height, in pixels. Raises +ArgumentError+ if +height+ is less than 1.
    def height=(height)
      raise ArgumentError, "Invalid height: #{height.inspect}" if height.nil? || height < 1
      @height = height
    end

    # Returns the chart's size as "WIDTHxHEIGHT".
    def size
      "#{width}x#{height}"
    end

    # Sets the chart's size as "WIDTHxHEIGHT".
    def size=(size)
      self.width, self.height = size.split("x").collect { |n| Integer(n) }
    end

    # Returns the chart's URL.
    def to_url
      query = query_params.collect { |k, v| "#{k}=#{URI.escape(v)}" }.join("&")
      "#{GChart::URL}?#{query}"
    end

    # Returns the chart's generated PNG as a blob.
    def fetch
      open(to_url) { |io| io.read }
    end

    # Writes the chart's generated PNG to a file. If +io_or_file+ quacks like an IO,
    # +write+ will be called on it instead.
    def write(io_or_file="chart.png")
      return io_or_file.write(fetch) if io_or_file.respond_to?(:write)
      open(io_or_file, "w+") { |io| io.write(fetch) }
    end

    protected
    
    def query_params
      params = { "cht" => render_chart_type, "chs" => size }
    
      render_data(params)
      render_title(params)
      render_colors(params)
      render_legend(params)

      params.merge(extras)
    end

    def render_chart_type
      raise NotImplementedError, "override in subclasses"
    end
    
    def render_data(params)
      raw = data && data.first.is_a?(Array) ? data : [data]
      max = self.max || raw.collect { |s| s.max }.max
  
      sets = raw.collect do |set|
        set.collect { |n| GChart.encode(:extended, n, max) }.join
      end
  
      params["chd"] = "e:#{sets.join(",")}"
    end
    
    def render_title(params)
      params["chtt"] = title.tr("\n", "|").gsub(/\s+/, "+") if title
    end
    
    def render_colors(params)
      params["chco"] = colors.join(",") if colors  
    end
    
    def render_legend(params)
      params["chdl"] = legend.join("|") if legend
    end    
  end
end
